[inductor] Do not reuse buffers across  scopes in mem planning

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120777
* #120665

Summary: Previously, in the `memory_plan_reuse` we assumed that the generated code is flat: in the sense of it can't have nested scopes. However, with nested control flow codegen-ing, this is no longer the case. This causes bugs in buffers being reused across the visibility boundaries in different nested scopes.

In this PR, we add nested planning states in `memory_plan_reuse` on entering and exiting scope in the codegen. This restricts the buffer reusability only to the currently active (peak) scope / planning state.

Test Plan:

```
python test/inductor/test_control_flow.py -k test_subgraphs_with_parameters
...
----------------------------------------------------------------------
Ran 27 tests in 149.413s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

aakhundov committed:
[inductor] Do not reuse buffers across  scopes in mem planning

Summary: Previously, in the `memory_plan_reuse` we assumed that the generated code is flat: in the sense of it can't have nested scopes. However, with nested control flow codegen-ing, this is no longer the case. This causes bugs in buffers being reused across the visibility boundaries in different nested scopes.

In this PR, we add nested planning states in `memory_plan_reuse` on entering and exiting scope in the codegen. This restricts the buffer reusability only to the currently active (peak) scope / planning state.

Test Plan:

```
python test/inductor/test_control_flow.py -k test_subgraphs_with_parameters
...
----------------------------------------------------------------------
Ran 27 tests in 149.413s

OK
```

[ghstack-poisoned]

aakhundov committed:
Update on "[inductor] Do not reuse buffers across  scopes in mem planning"


Summary: Previously, in the `memory_plan_reuse` we assumed that the generated code is flat: in the sense of it can't have nested scopes. However, with nested control flow codegen-ing, this is no longer the case. This causes bugs in buffers being reused across the visibility boundaries in different nested scopes.

In this PR, we add nested planning states in `memory_plan_reuse` on entering and exiting scope in the codegen. This restricts the buffer reusability only to the currently active (peak) scope / planning state.

Test Plan:

```
python test/inductor/test_control_flow.py -k test_subgraphs_with_parameters
...
----------------------------------------------------------------------
Ran 27 tests in 149.413s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -744,16 +744,28 @@ def memory_plan_reuse(self):
+                planning_states.append(MemoryPlanningState())
+            elif isinstance(line, ExitScopeLine):
+                past_planning_states.append(planning_states.pop())
+        past_planning_states.append(planning_states.pop())
```
nit - could we add an assert to ensure we have a single ```MemoryPlanningState``` instance after the loop? Just to make sure we have paired enter/exit states. Thanks.

chenyang78 reviewed (approved):
Left a minor comment. LGTM otherwise. Thanks!

aakhundov committed:
Update on "[inductor] Do not reuse buffers across  scopes in mem planning"


Summary: Previously, in the `memory_plan_reuse` we assumed that the generated code is flat: in the sense of it can't have nested scopes. However, with nested control flow codegen-ing, this is no longer the case. This causes bugs in buffers being reused across the visibility boundaries in different nested scopes.

In this PR, we add nested planning states in `memory_plan_reuse` on entering and exiting scope in the codegen. This restricts the buffer reusability only to the currently active (peak) scope / planning state.

Test Plan:

```
python test/inductor/test_control_flow.py -k test_subgraphs_with_parameters
...
----------------------------------------------------------------------
Ran 27 tests in 149.413s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1395,13 +1408,13 @@ def codegen_conditional(self, conditional):
         self.writeline(f"if {conditional.predicate.codegen_reference()}.item():")
-        self.writeline(IndentLine())
+        self.writeline(EnterScopeLine())
         self.codegen_subgraph(conditional.true_subgraph, outer_inputs, outer_outputs)
```
Question: Is there any benefit from intermingling the codegen of subgraph with the parent graph? Could we instead codegen  the subgraph independently and treat that function as an opaque operator in the parent graph?

aakhundov replied:
Great question! Codegening subgraphs in separate functions is, indeed, an alternative. However, when deciding between the two alternatives, I've taken the inlining approach in https://github.com/pytorch/pytorch/pull/119759 mainly for two reasons:

1. To make the symbols defined in the main `call` function (e.g., the dynamic shape symbols) readily available in the nested subgraph codegen. Deriving dynamic shape symbols is not necessarily possible from the subgraph inputs, as those can have expressions in the shapes. So we'd have to pass dynamic shape symbols, too.

2. The current structure of the wrapper codegen and the way it's utilized in the client code seemed non-trivial to adapt to generating subgraphs as functions away from the current "cursor" in the wrapper.

aakhundov commented:
@pytorchbot merge

izaitsevfb commented:
@aakhundov, looks like this PR is breaking ROCm ConTests:
```
FAILED [0.9116s] inductor/test_control_flow.py::CondTests::test_subgraphs_with_parameters_device_cuda_dynamic_True - AssertionError: Scalars are not equal!

Expected 1 but got 2.
Absolute difference: 1
Relative difference: 1.0
only one compilation expected
```
https://github.com/pytorch/pytorch/actions/runs/8090761332/job/22109394817#step:14:1011

aakhundov commented:
@izaitsevfb thanks for the heads up, taking a look!

aakhundov commented:
Discussed offline with @izaitsevfb: as the failing test was added in this PR, we disable it in https://github.com/pytorch/pytorch/issues/120955 pending further investigation on why it fails on AMD. Fixing the test should close https://github.com/pytorch/pytorch/issues/120955.
